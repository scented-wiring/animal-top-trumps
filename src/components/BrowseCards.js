import "../styles/BrowseCards.css";
import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import axios from "axios";
import Alert from "../components/Alert";
import Card from "../components/Card";

const BrowseCards = () => {
  const initialState = {
    alert: {
      message: "",
      alertType: "",
    },
    card: {
      name: "Choose a card from the left to view its stats",
      aka: "default__card",
    },
  };

  const [cards, setCards] = useState([]);
  const [card, setCard] = useState(initialState.card);
  const [alert, setAlert] = useState(initialState.alert);
  const [load, setLoad] = useState(true);

  useEffect(() => {
    setLoad(true);
    axios
      .get("http://localhost:3000/cards")
      .then((response) => {
        setCards(response.data);
        setLoad(false);
      })
      .catch(() => {
        setAlert({
          message: "Could not connect to the server.",
          alertType: "Error",
        });
        setLoad(false);
      });
  }, []);

  const handleDeleteCard = (id) => {
    axios
      .delete(`http://localhost:3000/cards/${id}`)
      .then(() => setCards(cards.filter((card) => card.id !== id)))
      .then(() => {
        setCard(initialState.card);
      })
      .then(() => {
        setAlert({
          message: "Card deleted",
          alertType: "Success",
        });
      })
      .catch(() => {
        setAlert({
          message: "Could not connect to the server",
          alertType: "Error",
        });
      });
  };

  const handleAlertPress = () => {
    setAlert({
      message: "",
      alertType: "",
    });
  };

  if (!load) {
    return (
      <div className="BrowseCards">
        <h2>Browse Cards</h2>
        <div className="display">
          <select
            name="selectCards"
            id="selectCards"
            size="15"
            onChange={(e) =>
              setCard(
                cards[document.getElementById("selectCards").selectedIndex]
              )
            }
          >
            {cards.map((card) => (
              <option key={cards.indexOf(card)}>{card.name}</option>
            ))}
          </select>
          {!cards.length ? (
            <Card
              name={`No cards found! Go to "Create Cards" from the menu to add cards to your deck.`}
              aka="default__card"
            />
          ) : (
            <Card {...card} deleteCard={handleDeleteCard} />
          )}
        </div>
        <Alert
          message={alert.message}
          alertType={alert.alertType}
          onAlertPress={handleAlertPress}
        />
        <Link to="/">
          <h3>&larr; BACK</h3>
        </Link>
      </div>
    );
  } else {
    return (
      <div className="BrowseCards">
        <h2>Loading...</h2>
      </div>
    );
  }
};

export default BrowseCards;
