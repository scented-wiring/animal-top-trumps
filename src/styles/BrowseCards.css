.BrowseCards {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.display {
  display: flex;
}

#selectCards {
  margin-right: 30px;
  align-self: center;
  font-size: 23px;
  font-family: "Montserrat", sans-serif;
  height: 460px;
  width: 230px;
  border: 2px black solid;
  box-shadow: 10px 10px #004742;
}

#selectCards:focus {
  outline: 0;
}

#selectCards-drop {
  display: none;
}

#delete {
  width: 170px;
  margin: 25px 10px 0 10px;
}

@media only screen and (max-width: 625px) {
  #selectCards {
    font-size: 20px;
    height: 400px;
    width: 200px;
    box-shadow: 8px 8px #004742;
  }
}

@media only screen and (max-width: 525px) {
  .display {
    flex-direction: column;
  }

  #selectCards {
    display: none;
  }

  #selectCards-drop {
    display: inline;
    margin-bottom: 15px;
  }

  #selectCards-drop:focus {
    outline: none;
  }
}

@media only screen and (max-width: 380px) {
  #browse-buttons {
    display: flex;
    flex-direction: column;
  }
}
