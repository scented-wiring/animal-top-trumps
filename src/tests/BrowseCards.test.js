import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import BrowseCards from "../components/BrowseCards";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { fireEvent, getByText, findByText } from "@testing-library/react";

const renderContainer = () => {
  render(
    <BrowserRouter>
      <BrowseCards />
    </BrowserRouter>,
    container
  );
};

const fakeResponse = [
  { id: 1, name: "Cat", aka: "Furry demon", alignment: "Lawful Evil" },
  { id: 2, name: "Dog" },
  { id: 3, name: "Elephant" },
];

let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("BrowseCards component", () => {
  test("renders correctly", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });
    expect(container).toMatchSnapshot();
  });

  test("renders a loading message", async () => {
    renderContainer();

    expect(await findByText(container, "Loading...")).toBeInTheDocument();
  });

  test("renders an alert if API can't connect to server", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(404);

    await act(async () => {
      renderContainer();
    });

    expect(container.textContent).toContain(
      "ErrorCould not connect to the server"
    );
  });

  test("renders a message if no cards found", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, []);

    await act(async () => {
      renderContainer();
    });

    expect(container.textContent).toContain("No cards found!");
  });

  test("renders a page title", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });

    expect(container.textContent).toContain("Browse Cards");
  });

  test("renders an option for each object fetched from the API", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });

    const options = container.getElementsByTagName("option");

    expect(container.textContent).toContain("CatDogElephant");
    expect(options).toHaveLength(3);
  });

  test("renders a card with default text on load", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });
    expect(container.textContent).toContain(
      "Choose a card from the left to view its stats"
    );
  });

  test("shows stats of card from select menu when clicked", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);

    await act(async () => {
      renderContainer();
    });

    fireEvent.change(document.getElementById("selectCards"), {
      target: { value: "Cat" },
    });

    expect(container.textContent).toContain(
      'aka: "Furry demon" Alignment:Lawful Evil'
    );
  });

  test("renders an alert on 'delete all cards' press", async () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:3000/cards").reply(200, fakeResponse);
    mock.onDelete("http://localhost:3000/cards").reply(200);

    await act(async () => {
      renderContainer();
    });

    fireEvent.click(getByText(container, "DELETE ALL CARDS"));

    expect(container.textContent).toContain("Are you sure?");
  });
});
