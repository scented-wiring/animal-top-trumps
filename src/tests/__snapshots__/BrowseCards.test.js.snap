// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowseCards component renders correctly 1`] = `
<div>
  <div
    class="BrowseCards"
  >
    <h2>
      Browse Cards
    </h2>
    <div
      class="display"
    >
      <select
        class="selectCards"
        name="selectCards"
        size="15"
      >
        <option>
          Cat
        </option>
        <option>
          Dog
        </option>
        <option>
          Elephant
        </option>
      </select>
      <div
        class="Card-border"
      >
        <div
          class="Card"
        >
          <div
            class="other-stat"
          >
            Choose a card from the left to view its stats
          </div>
        </div>
      </div>
    </div>
    <a
      href="/"
    >
      <h3>
        ‚Üê BACK
      </h3>
    </a>
  </div>
</div>
`;
